### PR TITLE
Check for equality between Fortran and C logicals

### LINKED ISSUE #691: Error: Ambiguous interfaces in generic interface 'string_type' for 'new_string_from_logical_lk' at (1) and 'new_string_from_logical_c_bool' at (2)
### Description

```
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/build/src/stdlib_string_type.f90:74:8:

   74 |         elemental module function new_string_from_logical_lk(val) result(new)
      |        1
......
   78 |         elemental module function new_string_from_logical_c_bool(val) result(new)
      |        2
Error: Ambiguous interfaces in generic interface 'string_type' for 'new_string_from_logical_lk' at (1) and 'new_string_from_logical_c_bool' at (2)
make[2]: *** [src/CMakeFiles/fortran_stdlib.dir/stdlib_string_type.f90.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

### Expected Behaviour

Types should be consistent or at least not lead to build failure :)

### Version of stdlib

0.2.1

### Platform and Architecture

macOS 10.6.8/ppc

### Additional Information

gcc 12.2.0
cmake 3.25.2
python 3.11.1
py311-fypp 3.1

### PR DESCRIPTION
**Pull request by jvdp1**
Addition of CMake rules to check if Fortran and C logicals are equivalent. If so, code will not be compiled for `c_bool`

Fixes #691 

@barracuda156 Could you test this PR on your computer, please?

### CODE REVIEW

**Comment by barracuda156** [comment]
@jvdp1 Yes, sure. I will try in a while and update here. (Now the usable machine is building MPICH which takes forever.)
Thank you!

**Comment by barracuda156** [comment]
@jvdp1 I think that error got fixed with your patch, judging from percentage of the build completed.

Now it fails on:
```
:info:build /usr/bin/make  -f test/io/CMakeFiles/test_loadtxt_qp.dir/build.make test/io/CMakeFiles/test_loadtxt_qp.dir/depend
:info:build make[2]: Entering directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/build'
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:61: error: integer constant is too large for ‘long’ type
:info:build [ 23%] Generating test_loadtxt_qp.f90
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/build/test/io && /opt/local/bin/fypp-3.11 -DMAXRANK=4 -DWITH_CBOOL=0 -DWITH_QP=0 -DWITH_XDP=1 -DPROJECT_VERSION_MAJOR=0 -DPROJECT_VERSION_MINOR=2 -DPROJECT_VERSION_PATCH=1 -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/src /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/io/test_loadtxt_qp.fypp /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/build/test/io/test_loadtxt_qp.f90
:info:build /opt/local/bin/cmake -E cmake_copy_f90_mod test/bitsets/test_stdlib_bitset_large.mod test/bitsets/CMakeFiles/test_stdlib_bitset_large.dir/test_stdlib_bitset_large.mod.stamp GNU
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int read_keys()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:67: error: no matching function for call to ‘std::basic_ifstream<char, std::char_traits<char> >::basic_ifstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:465: note: candidates are: std::basic_ifstream<_CharT, _Traits>::basic_ifstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:451: note:                 std::basic_ifstream<_CharT, _Traits>::basic_ifstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:89: note:                 std::basic_ifstream<char, std::char_traits<char> >::basic_ifstream(const std::basic_ifstream<char, std::char_traits<char> >&)
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int write_nmhash32()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:81: error: no matching function for call to ‘std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:596: note: candidates are: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:580: note:                 std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:92: note:                 std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(const std::basic_ofstream<char, std::char_traits<char> >&)
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int write_nmhash32x()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:99: error: no matching function for call to ‘std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:596: note: candidates are: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:580: note:                 std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:92: note:                 std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(const std::basic_ofstream<char, std::char_traits<char> >&)
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int write_water()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:117: error: no matching function for call to ‘std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:596: note: candidates are: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:580: note:                 std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:92: note:                 std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(const std::basic_ofstream<char, std::char_traits<char> >&)
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int write_pengy()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:135: error: no matching function for call to ‘std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:596: note: candidates are: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:580: note:                 std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:92: note:                 std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(const std::basic_ofstream<char, std::char_traits<char> >&)
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp: In function ‘int write_spooky()’:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_fortran-stdlib/fortran-stdlib/work/stdlib-1243ad93cc3a6f228cbf9544e84aa0d9c14caeee/test/hash_functions/generate_hash_arrays.cpp:153: error: no matching function for call to ‘std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::string&, std::_Ios_Openmode)’
:info:build /usr/include/c++/4.2.1/fstream:596: note: candidates are: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::_Ios_Openmode) [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/fstream:580: note:                 std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char, _Traits = std::char_traits<char>]
:info:build /usr/include/c++/4.2.1/iosfwd:92: note:                 std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(const std::basic_ofstream<char, std::char_traits<char> >&)
:info:build /opt/local/bin/cmake -E cmake_copy_f90_mod test/io/test_loadtxt.mod test/io/CMakeFiles/test_loadtxt.dir/test_loadtxt.mod.stamp GNU
:info:build make[2]: *** [test/hash_functions/CMakeFiles/test_hash_functions.dir/generate_hash_arrays.cpp.o] Error 1
```

**Comment by barracuda156** [comment]
UPD. Oops, sorry, it picked wrong GCC. I will rerun with gcc12 now.

**Comment by barracuda156** [comment]
@jvdp1 Beautiful, it builds through completion now! Thank you!

P. S. That still requires fixes to GCC due to: https://github.com/fortran-lang/stdlib/issues/690

**Comment by jvdp1** [comment]
> @jvdp1 Beautiful, it builds through completion now! Thank you!

Thank you for confirming it.

@awvwgk Could you have a look at the changes and approve them if they are ok for you?

Also, there are an issue with MacOS9, but I can't understand what it is.


> P. S. That still requires fixes to GCC due to: #690

Ok. I will look at it.

**Comment by barracuda156** [comment]
> Also, there are an issue with MacOS9, but I can't understand what it is.

You do not refer to my reports here, right? Honestly I doubt anyone ever tried to build Fortran `stdlib` on macOS 9. Would be probably fun to try, but I have no environment for that.

> > P. S. That still requires fixes to GCC due to: #690
> 
> Ok. I will look at it.

Problem is discussed here: https://github.com/iains/darwin-toolchains-start-here/discussions/40
I have preliminary implementation here, but it needs more work: https://github.com/barracuda156/gcc-git/tree/gfortran

**Comment by awvwgk** [comment]
> Also, there are an issue with MacOS9, but I can't understand what it is.

Looks like our GCC 9 setup is currently broken. Here is the error message from the workflow:

```
     /usr/local/bin/gfortran-9 -fimplicit-none -ffree-line-length-132  -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.6 CMakeFiles/cmTC_971d0.dir/testFortranCompiler.f.o -o cmTC_971d0 
    gfortran-9: warning: '12.6' is not valid for 'mmacosx-version-min'
    
    ld: library not found for -lgcc_s.10.4
    collect2: error: ld returned 1 exit status
```

To me it looks like a two recent version is passed by CMake to the compiler, maybe inferred from `MACOSX_DEPLOYMENT_TARGET`?

File: config/CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,13 @@ set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}" PARENT_SCOPE)
 # Check for available features
 # Note: users can overwrite the automatic check by setting the value at configure time
 include(CheckFortranSourceRuns)
+if (NOT DEFINED WITH_CBOOL)
+  check_fortran_source_runs(
+    "use, intrinsic :: iso_c_binding, only: c_bool; integer, parameter :: lk = kind(.true.); if (c_bool == lk) stop 1; end"
```
Comment (awvwgk) [inline config/CMakeLists.txt:21]: This can fail due to line-length, better make it a multiline string.

**Comment by barracuda156** [comment]
> > Also, there are an issue with MacOS9, but I can't understand what it is.
> 
> Looks like our GCC 9 setup is currently broken. Here is the error message from the workflow:
> 
> ```
>      /usr/local/bin/gfortran-9 -fimplicit-none -ffree-line-length-132  -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.6 CMakeFiles/cmTC_971d0.dir/testFortranCompiler.f.o -o cmTC_971d0 
>     gfortran-9: warning: '12.6' is not valid for 'mmacosx-version-min'
>     
>     ld: library not found for -lgcc_s.10.4
>     collect2: error: ld returned 1 exit status
> ```
> 
> To me it looks like a too recent version is passed by CMake to the compiler, maybe inferred from `MACOSX_DEPLOYMENT_TARGET`?

Nothing beyond Tiger should link to that dylib, I believe. Weird that it is expecting it to be there.

**Comment by awvwgk** [comment]
I'm running the same GCC 9 version on another CI workflow, meson has no problem to use it for compiling Fortran code. Maybe there is an issue with the latest CMake in the compiler detection (wild guess)?

**Comment by barracuda156** [comment]
> I'm running the same GCC 9 version on another CI workflow, meson has no problem to use it for compiling Fortran code. Maybe there is an issue with the latest CMake in the compiler detection (wild guess)?

Having something like this broken for a new macOS would be strange, since it is widely tested and used. But I don’t have ready explanations.

File: config/CMakeLists.txt (hunk 1 above)
Comment (jvdp1) [reply config/CMakeLists.txt]: done

**Comment by jvdp1** [comment]
@awvwgk @barracuda156 I finally implemented @awvwgk 's request. Could you review this PR and merge it, if ok, please?

**Review by awvwgk — APPROVED** [review]
Looks good to me now. Feel free to merge once we can confirm it fixes the issue.

**Comment by barracuda156** [comment]
> Looks good to me now. Feel free to merge once we can confirm it fixes the issue.

It does fix the build, and we are already using the patch in Macports: https://github.com/macports/macports-ports/blob/master/lang/fortran-stdlib/files/0001-Check-for-equality-between-Fortran-and-C-logicals.patch
So yes, merging is welcomed.

**Pull request merged**